README.md
# wkdmod is a fork of DevilutionX
## wkdmod aims to implement more quality of life changes that are deemed out of scope for DevX.  This modified version of DevilutionX focuses on a more modern and casual approach to ARPG gameplay and mechanics.

**[CREDITS]**

This mod was created by <a href="https://www.patreon.com/wkd1337gmr/membership" target="_blank" rel="noopener noreferrer">wkd1337gmr</a>.

Thank you to the invaluable efforts of the DevilutionX community going above and beyond whenever I needed assistance during my journey creating this mod!  The following handles contributed to my mod via either direct or indirect code consultation/contribution or game testing:

- Ajenbo
- staphen
- Qndel
- kphoeniX
- DakkJaniels
- ikonomov
- Knight of NSFW
- Brewbeck
- Entkoffeinator
- Tree
- Belix (wkdmod is inspired by the efforts of Hellfire modders that came before me)

(click the banner for current intro trailer)
<a href="https://www.youtube.com/watch?v=qGFhbjFcxxQ" target="_blank" rel="noopener noreferrer">
<img src="https://github.com/wkdgmr/wkdmod/blob/1e45674d564003a906c145185e141bda8ed285d9/wkdmodbanner.jpg" alt="wkdmod banner" width="838">
</a>

<a href="https://www.youtube.com/watch?v=6lJHouylAdA" target="_blank" rel="noopener noreferrer">wkdmod gameplay trailer</a>

wkdmod is compatible* with standard DevilutionX characters. Simply copy your character data and stash data into the **devxmods/wkdmod folder in %appdata%/roaming/**

***WARNING**
1) Unique Items:  All Unique Items are now available and have been extensively overhauled.  Ice Shank and Hammer of Johlm have been replaced with the Amulet of Empowerment and Tsang's Blessed Hammer. **(Unique's are not forwards or backwards compatible from standard devx/vanilla characters.  If brought into or out of wkdmod they could cause game crashing/character save corruption).**  If you plan on bringing over your character and stash, make sure you make a copy of this stuff and before bringing in the character load up the save and toss out all of your uniques.  There have been many new uniques and effects added.  Play the game to discover them all!

2) Magic/Blue Items:  In Hell the player will be able to find all affixes in the dungeon.  There are no longer any items that "can't" be found. **(Magic/Blue Items are intended to be forwards compatible into wkdmod.  There may still be potentially a small percentage of these items that could morph... it's just the nature of what happens when you bring stuff into a mod)**

3) Oils and Scrolls are **NOT** forwards compatible into wkdmod.

So in summary, what can be expected to be forwards compatible are: 
**--** Character saves with only blue items
**--** Stashes with gold, blue items, elixirs, books.

Don't forget, you can always start fresh as well if you'd like!

Please report bugs in the #wkdmod channel in the DevilutionX discord, underneath the wkdmod bug reports thread.  Or use github bug reports on the wkdmod github repo.

### **DevX Release 1.5.x** with additional **NEW** features:

**[GLOBAL]**

- When Diablo is killed the game does not end and he drops loot (drops 6 items).
- Buffed loot drop chances 300% in Hell, 200% in Nightmare, Normal is unchanged.
- When killing Diablo, Nakrul, Hork Demon or Defiler loot drop tiers are determined by difficulty.  Best drops from these bosses will be found on Hell difficulty.
- Diablo and Nakrul have been buffed and are much more difficult to defeat.

- All magic affixes including the previous unfindable ones are now available to find in the dungeon
- All unique items can now be found (including brand new never before seen uniques and unique item effects)

- Nightmare and Hell monster HP/AC/ToHit has been increased
- previous Triple Immunes now have RESISTANCE TO MAGIC and can be hit by magic spells
- Nightmare and Hell difficulties have been buffed significantly

- MLVL's have been increased in Nightmare and Hell to allow for slightly better affix pools in the higher difficulties.
- Wirt's shop price cap has been raised from 200,000 to 300,000. 
- Items are no longer destroyed at 0 durability, they become disabled and you must repair.
- Players do not drop items on death, unless you enable **INVADER** mode, more details below.  -1% TOTAL CHARACTER Experience upon death.

- Shared XP in multiplayer games.
- Character's now have a natural random chance to dodge incoming missiles.  This scales with character level.
- Character's now gain a passive spell dodge affect based on resistances to that spell element.

- Black Deaths have a decreasing chance based on character level to steal 1 Vit point from the player, and they have been given the same LIFESTEAL mechanic as the Skeleton King.
- **CRIPPLING SHRINES EFFECTS HAVE BEEN REMOVED:**
  - shrines that increase spell levels and did -10% mana now do -2 to the magic stat
  - Spiritual Shrines gold pile amounts per difficulty are increase by x 20
  - Stone Shrine now increases equipped Staff Charges by a random amount based on character level as well as refilling them
  - Gloomy Shrine can't increase AC past 90 (balance fix for wkdmod)
  - Weird Shrine cannot increase weapon max damage past weapon maximums set in wkdmod, will also increase elemental damage if present
  **- HELLFIRE Shrines**
  - Glowing Shrine now increases magic randomly between 1 - 5 and gives XP between 1000 - 5000, removed penalty
  - Murphy's Shrine now has a 50% chance to increase a random items durability using the Oil of Fortitude logic and give you money using the Spiritual  Shrine logic, the downsides remain the same
  - This should make it so wkdmod players no longer need to use the disable crippling shrines option as all negative effects are reversible with oils or elixirs (or even other shrines)

- Warrior, Barbarian, and Bard have all received a melee and ranged ToHit buff and are no longer desperate for Dexterity/ToHit gear.
- Barbarian no longer takes damage from the Rage skill, it will drain mana instead if the Barbarian has any
- Buffed Barbarian's passive bonuses to include 2H Axes, 2H Maces, Staffs, 2H Swords
- Buffed Bard's passive bonuses with dual wielding any combination of 1H weapons
- Bard is now buffed when wearing Unique, Medium, and Light Armor.  However receives a debuff when wearing non-Unique Heavy Armor (same as Monk).
- Golem has been buffed significantly and can no longer take Friendly Fire damage.
- Rogue gains +100 on Normal, +500 on Nightmare, or +1000 on Hell experience points when disarming a trap
- The Thorns affix has been reworked into Holy Fire.  Holy Fire damage can be boosted by elemental damage Oils and will do full additive damage from any equipped Holy Fire source to any Monster that melee hits the player (Holy Shock has replaced Flash, Exploding Bone Armor added as special item empowered effect).
- The Damage% line of affixes have been buffed, ending with Merciless which now can get up to 250% damage (151 - 250).
- The +Damage line of affixes have been buffed, ending with Slaughter, which now can get up to +50 damage (17 - 50).
- all Lifesteal and Manasteal affixes are now additive
- Nerfed Apocalypse in Diablo so it acts like its Hellfire counterpart
- Swapped some Scrolls out for the Oils I added, nobody will miss them

**UNIQUE ITEM OVERHAUL**
- Quest uniques have been brought in line with how easy they are to obtain, but mostly have also received buffs (a small handful nerfed) and can be farmed for better/preferred rolls on stats
- Some items will seem to have more punishing curses on them, but this is only because of how powerful they are and the curses are really there to make it a dangerous item for under leveled characters
- Some Uniques are simply just going to be better than others, no way to avoid that.  But hopefully everything is at least useful if found in a playthrough even if not the most optimal piece in the long run... also trying to maintain theme of the items as well
- There have been a ton of unique items given spell-on-strike and bows that shoot spells... explore the dungeon to find these new powerful weapons!
- - New Item effect: Empowerment (Can be found on the Amulet of Empowerment (Hellfire) or the Amulet of the Horadrim (Diablo))
Many unique items have new EMPOWERED effects when combined with either of the two new powerful amulets waiting for you to discover them!

**SPELL AND ELEMENTAL WEAPON DAMAGE BALANCING**
- Holy Bolt is now Magic Missile
- Flash is now Holy Shock (even though its one of the new auto cast effects, it can still be manually cast as well)
- Magic Missile, Blood Star and Bone Spirit have had damage buffs in line with their balance in the game and Blood Star/Bone Spirit should be considered equivalent damage to Fireball
    - Bone Spirit being a significantly stronger spell for high magic users and it is more then triple the mana cost of Blood Star) (edited)
- Firebolt (+Guardian), Charged Bolt, Inferno and Lightning have had slight damage buffs to scale with character level, and these spells pierce enemy immunes and damage is resisted instead of ignored.
- Lightning and Fire WEAPON damage pierce enemy immunes and damage is resisted instead of ignored.  This includes the special Hellfire weapons that cast spells based on Weapon element damage.
- Mana shield has been rebalanced as well and no longer redirects 100% of damage to mana

**OILS**
- All available Oils now drop in Classic and Hellfire multiplayer mode as well as singleplayer mode and added the following NEW Oils with NEW changes to 
  Accuracy, Mastery, Hardening, Imperviousness, Sharpening, Death, Permanence (note Blacksmith Oil has been removed):
- Oil of Sharpening:  +1 minimum dmg, +2 maximum damage, with 15/20/30/40 being the ceiling for minimum dmg for bows/1h/2h/Axe respectively / Can never exceed Maximum Damage

docs/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## DevilutionX 1.5.1

### Features

#### Multiplayer

- Validate player stats
- Update the pvp arenas
- Rename "Loopback" to "Offline"

#### Stability / Performance / System

- Move hp/mana display and item graphics to gameplay options
- Validate properties when reloading items
- Demomode: Improve replay stability
- Update [Discord link](https://discord.gg/devilutionx)
- Display save game confirmation
- Reduce ram usage

#### Translations

- Update Simplified Chinese translation
- Update French translation
- Update German translation
- Update Greek translation
- Update Japanese translation
- Update Korean translation
- Update Portuguese translation
- Update Spanish translation
- Update Swedish translation
- Update Ukrainian translation

### Bugfixes

#### Gameplay

- Being able to enter Lazarus' chamber before opening the portal
- Book requirements not updating
- Some monsters not walking
- Missiles not traveling the full distance at some angles
- Diablo: Incorrect level 4 layout when the Magic Banner quest is active
- Halls of the Blind not being completed by picking up the amulet
- Shareware: Bucklers not dropping
- Player animation stuttering

#### Multiplayer

- Potions dropped by Divine shrines not being synced

#### Platforms

- Linux: Add sdl-image dependency for the deb package
- Linux: Include discord dependency
- Xbox One: Missing assets

#### Graphics / Audio

- Incorrect cursor rendering
- Incorrect outlines at the right edge of the screen
- NPC speech continuing after starting a new game
- Correct various font rendering issues
- Hide the hit indicator when only one player is in the game
- Issues with flashing lights
- Floating number still appearing after death
- Misaligned automap

#### Controls

- Inconsistencies with placing items in to the stash
- Gamepad: Being stuck in dialogs
- Gamepad: Unable to use some scrolls directly

#### Stability / Performance / System

- Unable to playback new demo files
- Various crashes

### Bugfixes for original Diablo bugs

#### Gameplay

- Durability overflowing when reloading items
- Teleporting onto an occupied tile
- Right-click during dialogs casts spells

#### Graphics / Audio

- Cursor jitter when interacting with the inventory
- Broken lava tiles

#### Controls

- Inconsistencies with placing items in to the inventory

### Bugfixes for original Hellfire bugs

#### Gameplay

- Warping onto a solid tile

## DevilutionX 1.5.0

### Features

#### Gameplay

- Floating damage numbers
- Option to auto-pick up oils
- Quest items now drop based on difficulty

#### Multiplayer

- All quests can now be played, enabled from settings menu
- Add chat commands, use `/help` for options
- Add PvP arenas that can be accessed via `/arena #`
- Inspect other players' items using `/inspect <name>`
- Hellfire: Disabled The Cornerstone of the World
- Draw hostile players' names in red in the chat panel

#### Platforms

- Added support for the original Xbox
- Added support for Android TV
- Added (experimental) support for RG99
- Provide virtual resolutions for systems with only one native resolution
- Android: Support loading data from external storage
- Android: Do not speed up game after suspending if playing alone
- Windows: Digitally signed executable
- Support demo playback on SDL1 versions

#### Graphics / Audio

- Option to show item graphics in stores
- More fluid lighting updates when moving
- Added map in town
- Add widescreen load screens (thanks to @Chance4us)
- Durability icon gradually changing to red
- Provide sound cue when gold is auto-placed into the open inventory
- Color inventory slot based on item quality
- Show cast animation in town
- Monsters end their invisibility on death
- Make the ground color of small rejuvenation potions distinct
- Add setting for choosing a different audio resampler
- Support for recoloring heroes using trn-files
- Indicate on which difficulty a hero has defeated Diablo
- Use decimal separators for gold and XP
- Improve the monster HP bar for some edge cases
- Only display resolution height when FitToScreen is enabled
- Various text rendering improvements

#### Controls

- Do not target monster when casting Heal Other and Resurrect
- Improve logic for belt auto-refill
- Adjust the info panel descriptions based on the input device
- keyboard/mouse: Additional mapping options
- keyboard/mouse: Support back button in menus
- gamepad/touch: Improve menu navigation
- gamepad: Controls can now be mapped
- gamepad: Autodetect button layout

#### Translations

- Update French translation
- Update Italian translation
- Update Japanese translation
- Update Polish translation
- Update Spanish translation

Source/itemdat.h
/**
 * @file itemdat.h
 *
 * Interface of all item data.
 */
#pragma once

#include <cstdint>

#include "objdat.h"
#include "spelldat.h"
#include "utils/stdcompat/string_view.hpp"

namespace devilution {

/** @todo add missing values and apply */
enum _item_indexes : int16_t { // TODO defines all indexes in AllItemsList
	IDI_GOLD,
	IDI_WARRIOR,
	IDI_WARRSHLD,
	IDI_WARRCLUB,
	IDI_ROGUE,
	IDI_SORCERER,
	IDI_CLEAVER,
	IDI_FIRSTQUEST = IDI_CLEAVER,
	IDI_SKCROWN,
	IDI_INFRARING,
	IDI_ROCK,
	IDI_OPTAMULET,
	IDI_TRING,
	IDI_BANNER,
	IDI_HARCREST,
	IDI_STEELVEIL,
	IDI_GLDNELIX,
	IDI_ANVIL,
	IDI_MUSHROOM,
	IDI_BRAIN,
	IDI_FUNGALTM,
	IDI_SPECELIX,
	IDI_BLDSTONE,
	IDI_MAPOFDOOM,
	IDI_LASTQUEST = IDI_MAPOFDOOM,
	IDI_EAR,
	IDI_HEAL,
	IDI_MANA,
	IDI_IDENTIFY,
	IDI_PORTAL,
	IDI_ARMOFVAL,
	IDI_FULLHEAL,
	IDI_FULLMANA,
	IDI_GRISWOLD,
	IDI_LGTFORGE,
	IDI_LAZSTAFF,
	IDI_RESURRECT,
	IDI_OIL,
	IDI_SHORTSTAFF,
	IDI_BARDSWORD,
	IDI_BARDDAGGER,
	IDI_RUNEBOMB,
	IDI_THEODORE,
	IDI_AURIC,
	IDI_NOTE1,
	IDI_NOTE2,
	IDI_NOTE3,
	IDI_FULLNOTE,
	IDI_BROWNSUIT,
	IDI_GREYSUIT,
	IDI_BOOK1 = 114,
	IDI_BOOK2,
	IDI_BOOK3,
	IDI_BOOK4,
	IDI_BARBARIAN = 139,
	IDI_RUNEOFSTONE = 165,
	IDI_SORCERER_DIABLO,
	IDI_ARENAPOT,
	IDI_OILDEBUG,

	IDI_LAST = IDI_OILDEBUG,
	IDI_NONE = -1,
};

enum item_drop_rate : uint8_t {
	IDROP_NEVER,
	IDROP_REGULAR,
	IDROP_DOUBLE,
};

enum item_class : uint8_t {
	ICLASS_NONE,
	ICLASS_WEAPON,
	ICLASS_ARMOR,
	ICLASS_MISC,
	ICLASS_GOLD,
	ICLASS_QUEST,
};

enum item_equip_type : int8_t {
	ILOC_NONE,
	ILOC_ONEHAND,
	ILOC_TWOHAND,
	ILOC_ARMOR,
	ILOC_HELM,
	ILOC_RING,
	ILOC_AMULET,
	ILOC_UNEQUIPABLE,
	ILOC_BELT,
	ILOC_INVALID = -1,
};

/// Item graphic IDs; frame_num-11 of objcurs.cel.
enum item_cursor_graphic : uint8_t {
	// clang-format off
	ICURS_POTION_OF_FULL_MANA         = 0,
	ICURS_SCROLL_OF                   = 1,
	ICURS_GOLD_SMALL                  = 4,
	ICURS_GOLD_MEDIUM                 = 5,
	ICURS_GOLD_LARGE                  = 6,
	ICURS_RING_OF_TRUTH               = 10,
	ICURS_RING                        = 12,
	ICURS_SPECTRAL_ELIXIR             = 15,
	ICURS_ARENA_POTION                = 16,
	ICURS_GOLDEN_ELIXIR               = 17,
	ICURS_EMPYREAN_BAND               = 18,
	ICURS_EAR_SORCERER                = 19,
	ICURS_EAR_WARRIOR                 = 20,
	ICURS_EAR_ROGUE                   = 21,
	ICURS_BLOOD_STONE                 = 25,
	ICURS_OIL                         = 30,
	ICURS_ELIXIR_OF_VITALITY          = 31,
	ICURS_POTION_OF_HEALING           = 32,
	ICURS_POTION_OF_FULL_REJUVENATION = 33,
	ICURS_ELIXIR_OF_MAGIC             = 34,
	ICURS_POTION_OF_FULL_HEALING      = 35,
	ICURS_ELIXIR_OF_DEXTERITY         = 36,
	ICURS_POTION_OF_REJUVENATION      = 37,
	ICURS_ELIXIR_OF_STRENGTH          = 38,
	ICURS_POTION_OF_MANA              = 39,
	ICURS_BRAIN                       = 40,
	ICURS_OPTIC_AMULET                = 44,
	ICURS_AMULET                      = 45,
	ICURS_DAGGER                      = 51,
	ICURS_BLADE                       = 56,
	ICURS_BASTARD_SWORD               = 57,
	ICURS_MACE                        = 59,
	ICURS_LONG_SWORD                  = 60,
	ICURS_BROAD_SWORD                 = 61,
	ICURS_FALCHION                    = 62,
	ICURS_MORNING_STAR                = 63,
	ICURS_SHORT_SWORD                 = 64,
	ICURS_CLAYMORE                    = 65,
	ICURS_CLUB                        = 66,
	ICURS_SABRE                       = 67,
	ICURS_SPIKED_CLUB                 = 70,
	ICURS_SCIMITAR                    = 72,
	ICURS_FULL_HELM                   = 75,
	ICURS_MAGIC_ROCK                  = 76,
	ICURS_THE_UNDEAD_CROWN            = 78,
	ICURS_HELM                        = 82,
	ICURS_BUCKLER                     = 83,
	ICURS_VIEL_OF_STEEL               = 85,
	ICURS_BOOK_GREY                   = 86,
	ICURS_BOOK_RED                    = 87,
	ICURS_BOOK_BLUE                   = 88,
	ICURS_BLACK_MUSHROOM              = 89,
	ICURS_SKULL_CAP                   = 90,
	ICURS_CAP                         = 91,
	ICURS_HARLEQUIN_CREST             = 93,
	ICURS_CROWN                       = 95,
	ICURS_MAP_OF_THE_STARS            = 96,
	ICURS_FUNGAL_TOME                 = 97,
	ICURS_GREAT_HELM                  = 98,
	ICURS_BATTLE_AXE                  = 101,
	ICURS_HUNTERS_BOW                 = 102,
	ICURS_FIELD_PLATE                 = 103,
	ICURS_SMALL_SHIELD                = 105,
	ICURS_CLEAVER                     = 106,
	ICURS_STUDDED_LEATHER_ARMOR       = 107,
	ICURS_SHORT_STAFF                 = 109,
	ICURS_TWO_HANDED_SWORD            = 110,
	ICURS_CHAIN_MAIL                  = 111,
	ICURS_SMALL_AXE                   = 112,
	ICURS_KITE_SHIELD                 = 113,
	ICURS_SCALE_MAIL                  = 114,
	ICURS_SHORT_BOW                   = 118,
	ICURS_LONG_BATTLE_BOW             = 119,
	ICURS_LONG_WAR_BOW                = 120,
	ICURS_WAR_HAMMER                  = 121,
	ICURS_MAUL                        = 122,
	ICURS_LONG_STAFF                  = 123,
	ICURS_WAR_STAFF                   = 124,
	ICURS_TAVERN_SIGN                 = 126,
	ICURS_HARD_LEATHER_ARMOR          = 127,
	ICURS_RAGS                        = 128,
	ICURS_QUILTED_ARMOR               = 129,
	ICURS_FLAIL                       = 131,
	ICURS_TOWER_SHIELD                = 132,
	ICURS_COMPOSITE_BOW               = 133,
	ICURS_GREAT_SWORD                 = 134,
	ICURS_LEATHER_ARMOR               = 135,
	ICURS_SPLINT_MAIL                 = 136,
	ICURS_ROBE                        = 137,
	ICURS_ANVIL_OF_FURY               = 140,
	ICURS_BROAD_AXE                   = 141,
	ICURS_LARGE_AXE                   = 142,
	ICURS_GREAT_AXE                   = 143,
	ICURS_AXE                         = 144,
	ICURS_LARGE_SHIELD                = 147,
	ICURS_GOTHIC_SHIELD               = 148,
	ICURS_CLOAK                       = 149,
	ICURS_CAPE                        = 150,
	ICURS_FULL_PLATE_MAIL             = 151,
	ICURS_GOTHIC_PLATE                = 152,
	ICURS_BREAST_PLATE                = 153,
	ICURS_RING_MAIL                   = 154,
	ICURS_STAFF_OF_LAZARUS            = 155,
	ICURS_ARKAINES_VALOR              = 157,
	ICURS_SHORT_WAR_BOW               = 165,
	ICURS_COMPOSITE_STAFF             = 166,
	ICURS_SHORT_BATTLE_BOW            = 167,
	ICURS_GOLD                        = 168,
	ICURS_AURIC_AMULET                = 180,
	ICURS_RUNE_BOMB                   = 187,
	ICURS_THEODORE                    = 188,
	ICURS_TORN_NOTE_1                 = 189,
	ICURS_TORN_NOTE_2                 = 190,
	ICURS_TORN_NOTE_3                 = 191,
	ICURS_RECONSTRUCTED_NOTE          = 192,
	ICURS_RUNE_OF_FIRE                = 193,
	ICURS_GREATER_RUNE_OF_FIRE        = 194,
	ICURS_RUNE_OF_LIGHTNING           = 195,
	ICURS_GREATER_RUNE_OF_LIGHTNING   = 196,
	ICURS_RUNE_OF_STONE               = 197,
	ICURS_GREY_SUIT                   = 198,
	ICURS_BROWN_SUIT                  = 199,
	ICURS_BOVINE                      = 226,
	// clang-format on
};

enum class ItemType : int8_t {
	Misc,
	Sword,
	Axe,
	Bow,
	Mace,
	Shield,
	LightArmor,
	Helm,
	MediumArmor,
	HeavyArmor,
	Staff,
	Gold,
	Ring,
	Amulet,
	None = -1,
};

string_view ItemTypeToString(ItemType itemType);

enum unique_base_item : int8_t {
	UITYPE_NONE,
	UITYPE_SHORTBOW,
	UITYPE_LONGBOW,
	UITYPE_HUNTBOW,
	UITYPE_COMPBOW,
	UITYPE_WARBOW,
	UITYPE_BATTLEBOW,
	UITYPE_DAGGER,
	UITYPE_FALCHION,
	UITYPE_CLAYMORE,
	UITYPE_BROADSWR,
	UITYPE_SABRE,
	UITYPE_SCIMITAR,
	UITYPE_LONGSWR,
	UITYPE_BASTARDSWR,
	UITYPE_TWOHANDSWR,
	UITYPE_GREATSWR,
	UITYPE_CLEAVER,
	UITYPE_LARGEAXE,
	UITYPE_BROADAXE,
	UITYPE_SMALLAXE,
	UITYPE_BATTLEAXE,
	UITYPE_GREATAXE,
	UITYPE_MACE,
	UITYPE_MORNSTAR,
	UITYPE_SPIKCLUB,
	UITYPE_MAUL,
	UITYPE_WARHAMMER,
	UITYPE_FLAIL,
	UITYPE_LONGSTAFF,
	UITYPE_SHORTSTAFF,
	UITYPE_COMPSTAFF,
	UITYPE_QUARSTAFF,
	UITYPE_WARSTAFF,
	UITYPE_SKULLCAP,
	UITYPE_HELM,
	UITYPE_GREATHELM,
	UITYPE_CROWN,
	UITYPE_38,
	UITYPE_RAGS,
	UITYPE_STUDARMOR,
	UITYPE_CLOAK,
	UITYPE_ROBE,
	UITYPE_CHAINMAIL,
	UITYPE_LEATHARMOR,
	UITYPE_BREASTPLATE,
	UITYPE_CAPE,
	UITYPE_PLATEMAIL,
	UITYPE_FULLPLATE,
	UITYPE_BUCKLER,
	UITYPE_SMALLSHIELD,
	UITYPE_LARGESHIELD,
	UITYPE_KITESHIELD,
	UITYPE_GOTHSHIELD,
	UITYPE_RING,
	UITYPE_55,
	UITYPE_AMULET,
	UITYPE_SKCROWN,
	UITYPE_INFRARING,
	UITYPE_OPTAMULET,
	UITYPE_TRING,
	UITYPE_HARCREST,
	UITYPE_MAPOFDOOM,
	UITYPE_ELIXIR,
	UITYPE_ARMOFVAL,
	UITYPE_STEELVEIL,
	UITYPE_GRISWOLD,
	UITYPE_LGTFORGE,
	UITYPE_LAZSTAFF,
	UITYPE_BOVINE,
	UITYPE_INVALID = -1,
};

enum class ItemSpecialEffect : uint32_t {
	// clang-format off
	None                   = 0,
	RandomStealLife        = 1 << 1,
	RandomArrowVelocity    = 1 << 2,
	FireArrows             = 1 << 3,
	FireDamage             = 1 << 4,
	LightningDamage        = 1 << 5,
	DrainLife              = 1 << 6,
	DrainMana              = 1 << 7,
	Empower                = 1 << 8,
	MultipleArrows         = 1 << 9,
	Knockback              = 1 << 11,
	StealMana3             = 1 << 13,
	StealMana5             = 1 << 14,
	StealLife3             = 1 << 15,
	StealLife5             = 1 << 16,
	QuickAttack            = 1 << 17,
	FastAttack             = 1 << 18,
	FasterAttack           = 1 << 19,
	FastestAttack          = 1 << 20,
	FastHitRecovery        = 1 << 21,
	FasterHitRecovery      = 1 << 22,
	FastestHitRecovery     = 1 << 23,
	FastBlock              = 1 << 24,
	LightningArrows        = 1 << 25,
	Thorns                 = 1 << 26,
	NoMana                 = 1 << 27,
	HalfTrapDamage         = 1 << 28,
	TripleDemonDamage      = 1 << 30,
	MagicDamage            = 1U << 31,
	// clang-format on
};
use_enum_as_flags(ItemSpecialEffect);

enum class ItemSpecialEffectHf : uint16_t {
	// clang-format off
	None               = 0,
	Devastation        = 1 << 0,
	Decay              = 1 << 1,
	Peril              = 1 << 2,
	Jesters            = 1 << 3,
	Doppelganger       = 1 << 4,
	ACAgainstDemons    = 1 << 5,
	ACAgainstUndead    = 1 << 6,
	ManaToLife         = 1 << 7,
	LifeToMana         = 1 << 8,
	// clang-format on
};
use_enum_as_flags(ItemSpecialEffectHf);

enum item_misc_id : int8_t {
	IMISC_NONE,
	IMISC_USEFIRST,
	IMISC_FULLHEAL,
	IMISC_HEAL,
	IMISC_0x4, // Unused
	IMISC_0x5, // Unused
	IMISC_MANA,
	IMISC_FULLMANA,
	IMISC_0x8, // Unused
	IMISC_0x9, // Unused
	IMISC_ELIXSTR,
	IMISC_ELIXMAG,
	IMISC_ELIXDEX,
	IMISC_ELIXVIT,
	IMISC_0xE,  // Unused
	IMISC_0xF,  // Unused
	IMISC_0x10, // Unused
	IMISC_0x11, // Unused
	IMISC_REJUV,
	IMISC_FULLREJUV,
	IMISC_USELAST,
	IMISC_SCROLL,
	IMISC_SCROLLT,
	IMISC_STAFF,
	IMISC_BOOK,
	IMISC_RING,
	IMISC_AMULET,
	IMISC_UNIQUE,
	IMISC_0x1C, // Unused
	IMISC_OILFIRST,
	IMISC_OILOF, /* oils are beta or hellfire only */
	IMISC_OILACC,
	IMISC_OILDEATH,
	IMISC_OILSHARP,
	IMISC_OILMAST,
	IMISC_OILSKILL,
	IMISC_OILIMP,
	IMISC_OILPERM,
	IMISC_OILHARD,
	IMISC_OILFORT,
	IMISC_OILWICK,
	IMISC_OILFIRE,
	IMISC_OILDEBUG,
	IMISC_OILLAST,
	IMISC_MAPOFDOOM,
	IMISC_EAR,
	IMISC_SPECELIX,
	IMISC_0x2D, // Unused
	IMISC_RUNEFIRST,
	IMISC_RUNEF,
	IMISC_RUNEL,
	IMISC_GR_RUNEL,
	IMISC_GR_RUNEF,
	IMISC_RUNES,
	IMISC_RUNELAST,
	IMISC_AURIC,
	IMISC_NOTE,
	IMISC_ARENAPOT,
	IMISC_INVALID = -1,
};

struct ItemData {
	enum item_drop_rate iRnd;
	enum item_class iClass;
	enum item_equip_type iLoc;
	enum item_cursor_graphic iCurs;
	enum ItemType itype;
	enum unique_base_item iItemId;
	const char *iName;
	const char *iSName;
	uint8_t iMinMLvl;
	uint8_t iDurability;
	uint8_t iMinDam;
	uint8_t iMaxDam;
	uint8_t iMinAC;
	uint8_t iMaxAC;
	uint8_t iMinStr;
	uint8_t iMinMag;
	uint8_t iMinDex;
	ItemSpecialEffect iFlags; // ItemSpecialEffect as bit flags
	enum item_misc_id iMiscId;
	SpellID iSpell;
	bool iUsable;
	uint16_t iValue;
};

enum item_effect_type : int8_t {
	IPL_TOHIT,
	IPL_TOHIT_CURSE,
	IPL_DAMP,
	IPL_DAMP_CURSE,
	IPL_TOHIT_DAMP,
	IPL_TOHIT_DAMP_CURSE,
	IPL_ACP,
	IPL_ACP_CURSE,
	IPL_FIRERES,
	IPL_LIGHTRES,
	IPL_MAGICRES,
	IPL_ALLRES,
	IPL_SPLLVLADD = 14,
	IPL_CHARGES,
	IPL_FIREDAM,
	IPL_LIGHTDAM,
	IPL_MAGICDAM,
	IPL_STR = 19,
	IPL_STR_CURSE,
	IPL_MAG,
	IPL_MAG_CURSE,
	IPL_DEX,
	IPL_DEX_CURSE,
	IPL_VIT,
	IPL_VIT_CURSE,
	IPL_ATTRIBS,
	IPL_ATTRIBS_CURSE,
	IPL_GETHIT_CURSE,
	IPL_GETHIT,
	IPL_LIFE,
	IPL_LIFE_CURSE,
	IPL_MANA,
	IPL_MANA_CURSE,
	IPL_DUR,
	IPL_DUR_CURSE,
	IPL_INDESTRUCTIBLE,
	IPL_LIGHT,
	IPL_LIGHT_CURSE,
	IPL_MULT_ARROWS = 41, /* only used in hellfire */
	IPL_FIRE_ARROWS,
	IPL_LIGHT_ARROWS,
	IPL_INVCURS,
	IPL_THORNS,
	IPL_NOMANA,
	IPL_EMPOWER = 49,
	IPL_FIREBALL = 50, /* only used in hellfire */
	IPL_INFERNO,
	IPL_ABSHALFTRAP = 52,
	IPL_KNOCKBACK,
	IPL_BLOODSTARBOW,
	IPL_STEALMANA = 55,
	IPL_STEALLIFE,
	IPL_TARGAC,
	IPL_FASTATTACK,
	IPL_FASTRECOVER,
	IPL_FASTBLOCK,
	IPL_DAMMOD,
	IPL_RNDARROWVEL,
	IPL_SETDAM,
	IPL_SETDUR,
	IPL_NOMINSTR,
	IPL_SPELL,
	IPL_ONEHAND = 68,
	IPL_3XDAMVDEM,
	IPL_BONESPIRITBOW,
	IPL_ACIDBOW,
	IPL_DRAINLIFE = 72,
	IPL_RNDSTEALLIFE,
	IPL_DRAINMANA,
	IPL_SETAC = 75,
	IPL_ADDACLIFE,
	IPL_ADDMANAAC,
	IPL_HOLYBOLTBOW,
	IPL_AC_CURSE = 79,
	IPL_LASTDIABLO = IPL_AC_CURSE,
	IPL_FIRERES_CURSE,
	IPL_LIGHTRES_CURSE,
	IPL_MAGICRES_CURSE,
	IPL_DEVASTATION = 84,
	IPL_DECAY,
	IPL_PERIL,
	IPL_JESTERS,
	IPL_CRYSTALLINE,
	IPL_DOPPELGANGER,
	IPL_ACDEMON,
	IPL_ACUNDEAD,
	IPL_MANATOLIFE,
	IPL_LIFETOMANA,
	IPL_INVALID = -1,
};

enum goodorevil : uint8_t {
	GOE_ANY,
	GOE_EVIL,
	GOE_GOOD,
};

enum class AffixItemType : uint8_t {
	// clang-format off
	None      = 0,
	Misc      = 1 << 0,
	Bow       = 1 << 1,
	Staff     = 1 << 2,
	Weapon    = 1 << 3,
	Shield    = 1 << 4,
	Armor     = 1 << 5,
	// clang-format on
};
use_enum_as_flags(AffixItemType);

struct ItemPower {
	item_effect_type type = IPL_INVALID;
	int param1 = 0;
	int param2 = 0;
};

struct PLStruct {
	const char *PLName;
	ItemPower power;
	int8_t PLMinLvl;
	AffixItemType PLIType; // AffixItemType as bit flags
	enum goodorevil PLGOE;
	bool PLDouble;
	bool PLOk;
	int minVal;
	int maxVal;
	int multVal;
};

struct UniqueItem {
	const char *UIName;
	enum unique_base_item UIItemId;
	int8_t UIMinLvl;
	uint8_t UINumPL;
	int UIValue;
	ItemPower powers[6];
};

extern const ItemData AllItemsList[];
extern const PLStruct ItemPrefixes[];
extern const PLStruct ItemSuffixes[];
extern const UniqueItem UniqueItems[];

} // namespace devilution
